tech-specs/src/specs/ledger.md

```diff
@@ -0,0 +1,190 @@
+
+## The protocol
+
+- TODO describe DKG transactions
+
+### Transactions
+
+A transaction [encoded with proto3](./encoding.md#transactions) received from ABCI `DeliverTx` method is executed in two main steps:
+
+1. [Transaction execution](#transaction-execution)
+1. [Validity predicates check](#validity-predicates-check)
+
+#### Transaction execution
+
+For any error encountered in any of the following steps of transaction execution, the protocol MUST charge the gas used by the transaction and discard any storage changes that the transaction attempted to perform.
+
+1. Charge a base transaction [gas](#gas): \\[ \verb|BASE_TRANSACTION_FEE| \\]
+1. Decode the transaction bytes and validate the data.
+1. Charge WASM compilation gas, proportional to the bytes `length` of the `code` field of the transaction (this is because the WASM code is compiled with a single-pass compiler): \\[ \verb|length| * \verb|COMPILE_GAS_PER_BYTE| \\].
+1. [Validate the WASM code](#wasm-validation) from the `code` field of the transaction.
+1. Inject a [gas counter](#gas) into the `code`.
+1. Inject a [stack height](#stack-height-limiter) limiter into the `code`.
+1. Compile the transaction `code` with single-pass compiler. The compilation computational complexity MUST be linear in proportion to the `code` size.
+1. Initialize the WASM linear memory with descriptor having the initial memory size equal to [`TX_MEMORY_INIT_PAGES`](#wasm-constants) and maximum memory size to [`TX_MEMORY_MAX_PAGES`](#wasm-constants).
+1. Instantiate the WASM module with imported [transaction host environment functions](#transaction-host-environment-functions) and the instantiated WASM memory.
+1. Write the transaction's `data` into the memory exported from the WASM module instance.
+1. Attempt to call the module's entrypoint function. The entrypoint MUST have signature:
+   ```wat
+   func (param i64 i64)
+   ```
+   The first argument is the offset to the `data` input written into the memory and the second argument is its bytes length.
+
+If the transaction executed successfully, it is followed [Validity predicates check](#validity-predicates-check).
+
+#### Validity predicates check
+
+For the transaction to be valid, all the triggered validity predicates must accept it.
+
+First, the addresses whose validity predicates should be triggered by the transaction are determined. In this process, the addresses get associated with a set of modified storage keys that are relevant to the address:
+1. The addresses set by the transaction (see `insert_verifier` in [transaction host environment functions](#transaction-host-environment-functions)) are associated with *all* the modified storage keys.
+
+   TODO - <https://github.com/anomanetwork/anoma/issues/292>
+1. The storage keys that were modified by the transaction are associated with the addresses included in the storage keys. Note that a storage key may contain more than one address, in which case all its addresses are associated with this key. 
```

Comment:
did we decide to allow a storage key to be associated with more than one address? I thought we decided to make it possible for the executing code to trigger additional validity predicates